Проект ShareIt. Техническое задание. Часть 1

Akmula opened the pull request.

Akmula committed:
Проект ShareIt. Техническое задание. Часть 1

YazykovMikhail commented on src/main/java/ru/practicum/shareit/item/ItemController.java:
```diff
@@ -1,12 +1,64 @@
+    private final ItemService itemService;
+
+    @PostMapping
+    public ItemDto createItem(@RequestHeader("X-Sharer-User-Id") int userId,
```
Стоит разделять dto для запросов и ответов
ItemResponse - если все ответы одинаковые 

Или делать для каждого метода свое dto, если ответы разные, например
ItemCreateResponse
ItemUpdateResponse

аналогично и для запросов
Ведь например контракты дто для создание и обновления разные, при создании не нужны на пример поле id

Как итоге не должно быть лишних полей в дто,которые не используются в контракте конкретного метода

Akmula replied:
Сделал. Надеюсь все верно)

YazykovMikhail commented on src/main/java/ru/practicum/shareit/item/ItemDto.java:
```diff
@@ -0,0 +1,31 @@
+
+    @NotNull
+    @NotBlank
+    private String name;
```
Посмотри в чем разница между 
    @NotNull
    @NotBlank
и оставь что-то одно

Akmula replied:
Исправил

YazykovMikhail commented on src/main/java/ru/practicum/shareit/user/UserServiceImpl.java:
```diff
@@ -0,0 +1,81 @@
+@RequiredArgsConstructor
+public class UserServiceImpl implements UserService {
+
+    private Integer userId = 0;
```
генерация идентификаторов это ответсвенность хранилища, вынеси туда

Akmula replied:
Исправлено

YazykovMikhail commented on src/main/java/ru/practicum/shareit/user/UserRepositoryImpl.java:
```diff
@@ -0,0 +1,40 @@
+public class UserRepositoryImpl implements UserRepository {
+
+    private final Map<Integer, User> users = new HashMap<>();
+
```
Рекомендую сделать тут поле
	private final Set<String> emailUniqSet = new HashSet<>();
которое будет отвечать за уникальность имейлов, это будет такая же логика как ИНДЕКСы в БД (пройдешь в следюущем спринте)
Легко искать и проверять уникальность

Akmula replied:
Готово

YazykovMikhail commented on src/main/java/ru/practicum/shareit/user/UserRepositoryImpl.java:
```diff
@@ -0,0 +1,40 @@
+    private final Map<Integer, User> users = new HashMap<>();
+
+    @Override
+    public User addUser(User user) {
```
```
public User addUser(User user) {
		final String email = user.getEmail();
		if (emailUniqSet.contains(email)) {
			throw new DataAlreadyExistException("Email: " + email + " already exists");
		}
		user.setId(++counter);
		users.put(user.getId(), user);
		emailUniqSet.add(email);
		return user;
	}
	
```

Соответсвенно сохранение будет выглядеть как-то так

Akmula replied:
Сделано

YazykovMikhail commented on src/main/java/ru/practicum/shareit/user/UserRepositoryImpl.java:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    @Override
+    public User updateUser(User user) {
```
public void update(User user) {
		final String email = user.getEmail();
		users.computeIfPresent(user.getId(), (id, u) -> {
					if (!email.equals(u.getEmail())) {
						if (emailUniqSet.contains(email)) {
							throw new DataAlreadyExistException("Email: " + email + " already exists");
						}
						emailUniqSet.remove(u.getEmail());
						emailUniqSet.add(email);
					}
					return user;
				}
		);

	}

Akmula replied:
Исправил

Akmula commented:
Спасибо.

Akmula committed:
Проект ShareIt. Техническое задание. Часть 1
1. Изменить DTO в Item
2. Изменить аннотации у моделей
3. Генерацию идентификаторов перенести в хранилище
4. Сделать отдельное поле для хранения почты пользователей
5. Изменить методы при добавлении/обновлении пользователя

avfyodorov reviewed (commented):
Добрый вечер, Евгений!

Всё в порядке, замечаний нет. 
Работа принята.
